packages/firestore/exp/src/api/components.ts

```diff
@@ -16,138 +16,183 @@
-        'be enabled. You can only enable persistence before calling ' +
-        'any other methods on a Firestore object.'
+      'The client has already been terminated.'
     );
   }
-  logDebug(LOG_TAG, 'Initializing FirestoreClient');
-  const settings = firestore._getSettings();
-  const databaseInfo = new DatabaseInfo(
-    firestore._databaseId,
-    firestore._persistenceKey,
-    settings.host ?? DEFAULT_HOST,
-    settings.ssl ?? DEFAULT_SSL,
-    /** forceLongPolling= */ false
+}
+
+export function getSyncEngine(firestore: Firestore): Promise<SyncEngine> {
+  return getOnlineComponentProvider(firestore).then(
+    components => components.syncEngine
   );
-  const firestoreClient = new FirestoreClient(
-    firestore._credentials,
-    firestore._queue
+}
+
+export function getRemoteStore(firestore: Firestore): Promise<RemoteStore> {
+  return getOnlineComponentProvider(firestore).then(
+    components => components.remoteStore
   );
-  const initializationPromise = firestoreClient.start(
-    databaseInfo,
-    offlineComponentProvider,
-    onlineComponentProvider,
-    persistenceSettings
+}
+
+export function getEventManager(firestore: Firestore): Promise<EventManager> {
+  return getOnlineComponentProvider(firestore).then(
+    components => components.eventManager
   );
-  firestoreClientInstances.set(
-    firestore,
-    initializationPromise.then(() => firestoreClient)
+}
+
+export function getPersistence(firestore: Firestore): Promise<Persistence> {
+  return getOfflineComponentProvider(firestore).then(
+    components => components.persistence
   );
-  return initializationPromise;
 }
 
-/**
- * Removes and terminates the FirestoreClient for the given instance if it has
- * been started.
- */
-export async function removeFirestoreClient(
-  firestore: Firestore
-): Promise<void> {
-  const firestoreClientPromise = firestoreClientInstances.get(firestore);
-  if (firestoreClientPromise) {
-    firestoreClientInstances.delete(firestore);
-    return firestoreClientPromise.then(firestoreClient =>
-      firestoreClient.terminate()
-    );
-  }
+export function getLocalStore(firestore: Firestore): Promise<LocalStore> {
+  return getOfflineComponentProvider(firestore).then(
+    provider => provider.localStore
+  );
 }
 
 /**
  * Removes all components associated with the provided instance. Must be called
  * when the Firestore instance is terminated.
  */
 export async function removeComponents(firestore: Firestore): Promise<void> {
-  const firestoreClientPromise = firestoreClientInstances.get(firestore);
-  if (firestoreClientPromise) {
-    logDebug(LOG_TAG, 'Removing FirestoreClient');
-    firestoreClientInstances.delete(firestore);
-    return (await firestoreClientPromise).terminate();
+  const onlineComponentProviderPromise = onlineComponentProviders.get(
+    firestore
+  );
+  if (onlineComponentProviderPromise) {
+    logDebug(LOG_TAG, 'Removing OnlineComponentProvider');
```

Comment:
Question: nothing is logged if the promise was not found -- is this intentional?

Reply:
Yeah, I think that is ok. We only log when components are initialized and when they are terminated.